File: backend/tests/property_test_utils.py
"""
Property-Based Testing Utilities
Provides common strategies and helpers for Hypothesis tests
Requirements: 6.3 - Set up property-based testing
"""

from hypothesis import strategies as st, settings, Verbosity
from hypothesis.stateful import RuleBasedStateMachine, rule, invariant
from datetime import datetime, timedelta
import string


# ==================== Settings ====================

# Default settings for property tests
default_settings = settings(
    max_examples=100,  # Run 100 test cases
    verbosity=Verbosity.verbose,
    deadline=None,  # No deadline for slow tests
    print_blob=True  # Print failing examples
)

# Fast settings for quick tests
fast_settings = settings(
    max_examples=20,
    verbosity=Verbosity.normal,
    deadline=1000  # 1 second deadline
)

# Thorough settings for critical tests
thorough_settings = settings(
    max_examples=1000,
    verbosity=Verbosity.verbose,
    deadline=None
)


# ==================== Custom Strategies ====================

# Email addresses
def email_strategy():
    """Generate valid email addresses"""
    username = st.text(
        alphabet=string.ascii_lowercase + string.digits + '._-',
        min_size=1,
        max_size=64
    ).filter(lambda x: x[0] not in '.-' and x[-1] not in '.-')
    
    domain = st.text(
        alphabet=string.ascii_lowercase + string.digits + '-',
        min_size=1,
        max_size=63
    ).filter(lambda x: x[0] != '-' and x[-1] != '-')
    
    tld = st.sampled_from(['com', 'org', 'net', 'edu', 'gov'])
    
    return st.builds(
        lambda u, d, t: f"{u}@{d}.{t}",
        username, domain, tld
    )


# Passwords
def password_strategy(min_length=8, max_length=100):
    """Generate valid passwords"""
    return st.text(
        alphabet=string.ascii_letters + string.digits + string.punctuation,
        min_size=min_length,
        max_size=max_length
    )


# Topics
def topic_strategy(min_length=1, max_length=200):
    """Generate valid topic strings"""
    return st.text(
        alphabet=string.ascii_letters + string.digits + ' -.,()!?&',
        min_size=min_length,
        max_size=max_length
    ).filter(lambda x: x.strip() != '')


# Subtopics arrays
def subtopics_strategy(min_items=1, max_items=20):
    """Generate valid subtopics arrays"""
    return st.lists(
        topic_strategy(min_length=1, max_length=100),
        min_size=min_items,
        max_size=max_items,
        unique=True
    )


# JWT tokens
def jwt_token_strategy():
    """Generate JWT-like tokens"""
    header = st.text(alphabet=string.ascii_letters + string.digits, min_size=20, max_size=50)
    payload = st.text(alphabet=string.ascii_letters + string.digits, min_size=50, max_size=200)
    signature = st.text(alphabet=string.ascii_letters + string.digits, min_size=20, max_size=50)
    
    return st.builds(
        lambda h, p, s: f"{h}.{p}.{s}",
        header, payload, signature
    )


# API keys
def api_key_strategy(prefix='sk_'):
    """Generate API key strings"""
    key_part = st.text(
        alphabet=string.ascii_letters + string.digits,
        min_size=32,
        max_size=64
    )
    return st.builds(lambda k: f"{prefix}{k}", key_part)


# Timestamps
def timestamp_strategy(min_year=2020, max_year=2030):
    """Generate datetime objects"""
    return st.datetimes(
        min_value=datetime(min_year, 1, 1),
        max_value=datetime(max_year, 12, 31)
    )


# File sizes
def file_size_strategy(min_size=0, max_size=10*1024*1024):
    """Generate file sizes in bytes"""
    return st.integers(min_value=min_size, max_value=max_size)


# File extensions
def file_extension_strategy(allowed_only=True):
    """Generate file extensions"""
    if allowed_only:
        return st.sampled_from(['jpg', 'jpeg', 'png', 'gif', 'webp'])
    else:
        return st.text(
            alphabet=string.ascii_lowercase,
            min_size=2,
            max_size=5
        )


# HTTP status codes
def http_status_strategy():
    """Generate HTTP status codes"""
    return st.sampled_from([
        200, 201, 204,  # Success
        400, 401, 403, 404, 422, 429,  # Client errors
        500, 502, 503, 504  # Server errors
    ])


# User roles
def role_strategy():
    """Generate user roles"""
    return st.sampled_from(['user', 'admin', 'moderator'])


# Quota tiers
def quota_tier_strategy():
    """Generate quota tiers"""
    return st.sampled_from(['free', 'basic', 'premium', 'enterprise'])


# IP addresses
def ip_address_strategy():
    """Generate IP addresses"""
    octet = st.integers(min_value=0, max_value=255)
    return st.builds(
        lambda a, b, c, d: f"{a}.{b}.{c}.{d}",
        octet, octet, octet, octet
    )


# User agents
def user_agent_strategy():
    """Generate user agent strings"""
    browsers = ['Chrome', 'Firefox', 'Safari', 'Edge']
    versions = st.integers(min_value=80, max_value=120)
    
    return st.builds(
        lambda b, v: f"Mozilla/5.0 ({b}/{v})",
        st.sampled_from(browsers),
        versions
    )


# ==================== Composite Strategies ====================

def user_data_strategy():
    """Generate complete user data"""
    return st.fixed_dictionaries({
        'email': email_strategy(),
        'password': password_strategy(),
        'role': role_strategy(),
        'quota_tier': quota_tier_strategy(),
        'created_at': timestamp_strategy()
    })


def session_data_strategy():
    """Generate complete session data"""
    return st.fixed_dictionaries({
        'token': jwt_token_strategy(),
        'user_id': st.integers(min_value=1, max_value=1000000),
        'expires_at': timestamp_strategy(),
        'ip_address': ip_address_strategy(),
        'user_agent': user_agent_strategy()
    })


def api_key_data_strategy():
    """Generate complete API key data"""
    return st.fixed_dictionaries({
        'key': api_key_strategy(),
        'user_id': st.integers(min_value=1, max_value=1000000),
        'name': st.text(min_size=1, max_size=100),
        'created_at': timestamp_strategy()
    })


# ==================== Helper Functions ====================

def is_valid_email(email):
    """Check if email is valid"""
    import re
    pattern = r'^[a-zA-Z0-9._-]+@[a-zA-Z0-9-]+\.[a-zA-Z]{2,}$'
    return bool(re.match(pattern, email))


def is_valid_password(password, min_length=8):
    """Check if password meets requirements"""
    return len(password) >= min_length


def is_valid_topic(topic, max_length=200):
    """Check if topic is valid"""
    return 0 < len(topic.strip()) <= max_length


def is_valid_jwt(token):
    """Check if JWT format is valid"""
    parts = token.split('.')
    return len(parts) == 3 and all(len(p) > 0 for p in parts)


def is_valid_api_key(key, prefix='sk_'):
    """Check if API key format is valid"""
    return key.startswith(prefix) and len(key) > len(prefix) + 20


# ==================== State Machine Base ====================

class AuthStateMachine(RuleBasedStateMachine):
    """
    State machine for testing authentication flows
    Example of stateful property-based testing
    """
    
    def __init__(self):
        super().__init__()
        self.users = {}
        self.sessions = {}
        self.api_keys = {}
    
    @rule(email=email_strategy(), password=password_strategy())
    def register_user(self, email, password):
        """Register a new user"""
        if email not in self.users:
            self.users[email] = {
                'password': password,
                'role': 'user',
                'created_at': datetime.utcnow()
            }
    
    @rule(email=email_strategy(), password=password_strategy())
    def login_user(self, email, password):
        """Login a user"""
        if email in self.users and self.users[email]['password'] == password:
            token = f"token_{len(self.sessions)}"
            self.sessions[token] = {
                'email': email,
                'expires_at': datetime.utcnow() + timedelta(hours=24)
            }
    
    @invariant()
    def sessions_not_expired(self):
        """All sessions should not be expired"""
        now = datetime.utcnow()
        for token, session in self.sessions.items():
            # In real test, we'd check expiration
            pass


# ==================== Example Property Tests ====================

def example_password_property():
    """
    Example property test for password hashing
    This is a template - actual tests are in test files
    """
    from hypothesis import given
    
    @given(password=password_strategy())
    def test_password_hashing_roundtrip(password):
        # This would use actual PasswordHasher
        # hashed = hasher.hash(password)
        # assert hasher.verify(password, hashed)
        pass


def example_topic_property():
    """
    Example property test for topic validation
    This is a template - actual tests are in test files
    """
    from hypothesis import given
    
    @given(topic=topic_strategy())
    def test_topic_validation(topic):
        # This would use actual validator
        # result = validator.validate_topic(topic)
        # assert result.is_valid
        pass

File: frontend/src/__tests__/propertyTestUtils.js
/**
 * Property-Based Testing Utilities for Frontend
 * Provides common generators and helpers for fast-check tests
 * Requirements: 6.3 - Set up property-based testing
 */

import fc from 'fast-check';

// ==================== Custom Arbitraries ====================

/**
 * Generate valid email addresses
 */
export const emailArbitrary = () => {
  return fc.tuple(
    fc.stringOf(fc.constantFrom(...'abcdefghijklmnopqrstuvwxyz0123456789._-'), { minLength: 1, maxLength: 64 }),
    fc.stringOf(fc.constantFrom(...'abcdefghijklmnopqrstuvwxyz0123456789-'), { minLength: 1, maxLength: 63 }),
    fc.constantFrom('com', 'org', 'net', 'edu', 'gov')
  ).map(([user, domain, tld]) => `${user}@${domain}.${tld}`);
};

/**
 * Generate valid passwords
 */
export const passwordArbitrary = (minLength = 8, maxLength = 100) => {
  return fc.string({ minLength, maxLength });
};

/**
 * Generate valid topic strings
 */
export const topicArbitrary = (minLength = 1, maxLength = 200) => {
  const chars = 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789 -.,()!?&';
  return fc.stringOf(fc.constantFrom(...chars), { minLength, maxLength })
    .filter(s => s.trim().length > 0);
};

/**
 * Generate valid subtopics arrays
 */
export const subtopicsArbitrary = (minItems = 1, maxItems = 20) => {
  return fc.array(
    topicArbitrary(1, 100),
    { minLength: minItems, maxLength: maxItems }
  ).map(arr => [...new Set(arr)]); // Ensure unique
};

/**
 * Generate JWT-like tokens
 */
export const jwtTokenArbitrary = () => {
  return fc.tuple(
    fc.hexaString({ minLength: 20, maxLength: 50 }),
    fc.hexaString({ minLength: 50, maxLength: 200 }),
    fc.hexaString({ minLength: 20, maxLength: 50 })
  ).map(([header, payload, signature]) => `${header}.${payload}.${signature}`);
};

/**
 * Generate API keys
 */
export const apiKeyArbitrary = (prefix = 'sk_') => {
  return fc.hexaString({ minLength: 32, maxLength: 64 })
    .map(key => `${prefix}${key}`);
};

/**
 * Generate timestamps
 */
export const timestampArbitrary = () => {
  return fc.date({
    min: new Date('2020-01-01'),
    max: new Date('2030-12-31')
  });
};

/**
 * Generate file sizes in bytes
 */
export const fileSizeArbitrary = (minSize = 0, maxSize = 10 * 1024 * 1024) => {
  return fc.integer({ min: minSize, max: maxSize });
};

/**
 * Generate file extensions
 */
export const fileExtensionArbitrary = (allowedOnly = true) => {
  if (allowedOnly) {
    return fc.constantFrom('jpg', 'jpeg', 'png', 'gif', 'webp');
  }
  return fc.stringOf(fc.constantFrom(...'abcdefghijklmnopqrstuvwxyz'), { minLength: 2, maxLength: 5 });
};

/**
 * Generate HTTP status codes
 */
export const httpStatusArbitrary = () => {
  return fc.constantFrom(
    200, 201, 204,  // Success
    400, 401, 403, 404, 422, 429,  // Client errors
    500, 502, 503, 504  // Server errors
  );
};

/**
 * Generate user roles
 */
export const roleArbitrary = () => {
  return fc.constantFrom('user', 'admin', 'moderator');
};

/**
 * Generate quota tiers
 */
export const quotaTierArbitrary = () => {
  return fc.constantFrom('free', 'basic', 'premium', 'enterprise');
};

/**
 * Generate IP addresses
 */
export const ipAddressArbitrary = () => {
  return fc.tuple(
    fc.integer({ min: 0, max: 255 }),
    fc.integer({ min: 0, max: 255 }),
    fc.integer({ min: 0, max: 255 }),
    fc.integer({ min: 0, max: 255 })
  ).map(([a, b, c, d]) => `${a}.${b}.${c}.${d}`);
};

/**
 * Generate user agent strings
 */
export const userAgentArbitrary = () => {
  return fc.tuple(
    fc.constantFrom('Chrome', 'Firefox', 'Safari', 'Edge'),
    fc.integer({ min: 80, max: 120 })
  ).map(([browser, version]) => `Mozilla/5.0 (${browser}/${version})`);
};

// ==================== Composite Arbitraries ====================

/**
 * Generate complete user data
 */
export const userDataArbitrary = () => {
  return fc.record({
    email: emailArbitrary(),
    password: passwordArbitrary(),
    role: roleArbitrary(),
    quota_tier: quotaTierArbitrary(),
    created_at: timestampArbitrary()
  });
};

/**
 * Generate complete session data
 */
export const sessionDataArbitrary = () => {
  return fc.record({
    token: jwtTokenArbitrary(),
    user_id: fc.integer({ min: 1, max: 1000000 }),
    expires_at: timestampArbitrary(),
    ip_address: ipAddressArbitrary(),
    user_agent: userAgentArbitrary()
  });
};

/**
 * Generate complete API key data
 */
export const apiKeyDataArbitrary = () => {
  return fc.record({
    key: apiKeyArbitrary(),
    user_id: fc.integer({ min: 1, max: 1000000 }),
    name: fc.string({ minLength: 1, maxLength: 100 }),
    created_at: timestampArbitrary()
  });
};

/**
 * Generate error objects
 */
export const errorArbitrary = () => {
  return fc.record({
    error_code: fc.constantFrom(
      'VALIDATION_ERROR',
      'AUTH_REQUIRED',
      'RATE_LIMIT_EXCEEDED',
      'SERVER_ERROR'
    ),
    message: fc.string({ minLength: 10, maxLength: 200 }),
    details: fc.object()
  });
};

// ==================== Helper Functions ====================

/**
 * Check if email is valid
 */
export const isValidEmail = (email) => {
  const pattern = /^[a-zA-Z0-9._-]+@[a-zA-Z0-9-]+\.[a-zA-Z]{2,}$/;
  return pattern.test(email);
};

/**
 * Check if password meets requirements
 */
export const isValidPassword = (password, minLength = 8) => {
  return password.length >= minLength;
};

/**
 * Check if topic is valid
 */
export const isValidTopic = (topic, maxLength = 200) => {
  return topic.trim().length > 0 && topic.length <= maxLength;
};

/**
 * Check if JWT format is valid
 */
export const isValidJWT = (token) => {
  const parts = token.split('.');
  return parts.length === 3 && parts.every(p => p.length > 0);
};

/**
 * Check if API key format is valid
 */
export const isValidAPIKey = (key, prefix = 'sk_') => {
  return key.startsWith(prefix) && key.length > prefix.length + 20;
};

// ==================== Test Configuration ====================

/**
 * Default configuration for property tests
 */
export const defaultConfig = {
  numRuns: 100,
  verbose: true,
  seed: Date.now(),
};

/**
 * Fast configuration for quick tests
 */
export const fastConfig = {
  numRuns: 20,
  verbose: false,
};

/**
 * Thorough configuration for critical tests
 */
export const thoroughConfig = {
  numRuns: 1000,
  verbose: true,
};

// ==================== Example Property Tests ====================

/**
 * Example property test template
 */
export const examplePropertyTest = () => {
  // This is a template - actual tests are in test files
  
  test('password encryption is reversible', () => {
    fc.assert(
      fc.property(passwordArbitrary(), (password) => {
        // const encrypted = encrypt(password);
        // const decrypted = decrypt(encrypted);
        // return decrypted === password;
        return true;
      }),
      defaultConfig
    );
  });
};

/**
 * Example stateful test template
 */
export const exampleStatefulTest = () => {
  // This is a template for stateful testing
  
  class AuthModel {
    constructor() {
      this.users = new Map();
      this.sessions = new Map();
    }
    
    register(email, password) {
      if (!this.users.has(email)) {
        this.users.set(email, { password, role: 'user' });
        return true;
      }
      return false;
    }
    
    login(email, password) {
      const user = this.users.get(email);
      if (user && user.password === password) {
        const token = `token_${this.sessions.size}`;
        this.sessions.set(token, { email });
        return token;
      }
      return null;
    }
  }
  
  // Use fc.commands() for stateful testing
};

export default {
  // Arbitraries
  emailArbitrary,
  passwordArbitrary,
  topicArbitrary,
  subtopicsArbitrary,
  jwtTokenArbitrary,
  apiKeyArbitrary,
  timestampArbitrary,
  fileSizeArbitrary,
  fileExtensionArbitrary,
  httpStatusArbitrary,
  roleArbitrary,
  quotaTierArbitrary,
  ipAddressArbitrary,
  userAgentArbitrary,
  
  // Composite
  userDataArbitrary,
  sessionDataArbitrary,
  apiKeyDataArbitrary,
  errorArbitrary,
  
  // Helpers
  isValidEmail,
  isValidPassword,
  isValidTopic,
  isValidJWT,
  isValidAPIKey,
  
  // Config
  defaultConfig,
  fastConfig,
  thoroughConfig,
};
